wasmCloud/project-templates
# Added build and release actions for actor templates

This PR introduces two github actions, `build.yml` and `release.yml`. Build will build, test, lint, and check for formatting for an actor when PR-ed into `main`.

Release is where the fun part is, not only will it build and test the actor, but when you trigger the release action by pushing a tag the following will happen:
1. build, test, etc of actor
2. A Github release will be created including the OCI reference of that actor and the claims information
3. A Github package will be created and automatically associated with this repository, allowing you to start that actor from a remote wasmCloud host

The release action can also be triggered on a push to main or by a `workflow_dispatch` manually, either of which will publish to the `:latest` tag


If you'd like to give this a try for yourself, here's a draft of the instructions I plan to put in a blog post:
Instructions

1. `wash new actor --git https://github.com/brooksmtownsend/project-templates --branch feat/actor-actions --subfolder actor/hello hello`
2. create repo on github named `hello`
3. Run the following commands, substituting your github name
    
    ```bash
    cd hello
    git add .
    git commit -m "first commit"
    git branch -M main
    git remote add origin https://github.com/<your_github_name>/hello.git
    git push -u origin main
    ```
    
4. Adding secrets
    1. In your repository, navigate to the Settings tab and then to Secrets → Actions → New repository secret
        1. Run `wash keys gen account` and copy the 56 character `Seed` key, name that secret `WASH_ISSUER_KEY`(Note, you will not be able to access this secret again once you save it in Github, so save it in a safe place if you plan on signing this actor outside of Github in the future)
        2. Run `wash keys gen module` and copy the 56 character `Seed` key, name that secret `WASH_SUBJECT_KEY` (Same note as above regarding secret access)
        3. Follow [[Creating a personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token), ensuring you give the token `write:packages` access. Generate the token, it should be in the form of `ghp_xyz123....`. Copy that value and name a secret `WASMCLOUD_PAT`
5. Creating your initial release
    1. Back in the terminal, run the following commands to create and push a tag. This will kick off the release workflow, publishing the hello actor to GHCR and creating a github release for you.
    
    ```bash
    git tag -a v0.1.0 -m "initial release"
    git push -u origin v0.1.0
    ```
    
6. Making your actor public (optional)
    1. You can configure a wasmCloud host to pull your privately hosted actor, though it requires some additional configuration to authenticate.
    2. Navigate back to your repository and find the `Packages` section on the right-hand side. You should see a `hello` package once the release action finishes. Click on it
    3. On the right side click on `Package settings`
    4. Scroll to the bottom, click `Change visibility`, and then modify the visibility to `Public`. 

Now you can start your actor using the OCI reference found in your newly created release!
